src/main/scala/cognite/spark/v1/AssetsHierarchyBuilder.scala

```diff
@@ -0,0 +1,230 @@
+    extends Exception(message)
+final case class NoRootException(
+    message: String = """Tree has no root. Set parentExternalId to "" for the root Asset.""")
+    extends Exception(message)
+final case class InvalidTreeException(message: String = s"The tree is has an invalid structure.")
+    extends Exception(message)
+
+class AssetsHierarchyBuilder(config: RelationConfig)(val sqlContext: SQLContext)
+    extends CdfRelation(config, "assetshierarchybuilder") {
+
+  def build(df: DataFrame): IO[Unit] = {
+    val deleteMissingAssets = config.deleteMissingAssets
+    val batchSize = config.batchSize.getOrElse(Constants.DefaultBatchSize)
+
+    val sourceTree = df.collect.map(r => fromRow[AssetsIngestSchema](r))
+
+    val (root, children) = validateAndOrderInput(sourceTree)
+
+    for {
+      rootAndChildren <- getCdfRootAndTree(CogniteExternalId(root.externalId))
+      (cdfRoot, cdfChildren) = (rootAndChildren._1, rootAndChildren._2)
+      _ <- upsertRoot(root, cdfRoot)
+      (toDelete, toInsert, toUpdate) = nodesToDeleteInsertUpdate(root, children, cdfChildren)
+      _ <- insert(toInsert, batchSize)
+      _ <- update(toUpdate, batchSize)
+      _ <- delete(toDelete, deleteMissingAssets, batchSize)
+    } yield ()
+  }
+
+  def insert(toInsert: Seq[AssetsIngestSchema], batchSize: Int): IO[Unit] =
+    if (toInsert.nonEmpty) {
+      val assetCreatesToInsert = toInsert.map(AssetsIngestSchema.toAssetCreate).grouped(batchSize)
+      assetCreatesToInsert.toList.traverse(client.assets.create) *> IO.unit
+    } else {
+      IO.unit
+    }
+
+  def delete(toDelete: Seq[Asset], deleteMissingAssets: Boolean, batchSize: Int): IO[Unit] =
+    if (deleteMissingAssets) {
+      toDelete.flatMap(_.externalId) match {
+        case Nil => IO.unit
+        case externalIds =>
+          externalIds.grouped(batchSize).toList.traverse(client.assets.deleteByExternalIds) *> IO.unit
+      }
+    } else {
+      IO.unit
+    }
+
+  def update(toUpdate: Seq[AssetsIngestSchema], batchSize: Int): IO[Unit] =
+    if (toUpdate.nonEmpty) {
+      toUpdate
+        .map(a => a.externalId -> AssetsIngestSchema.toAssetUpdate(a))
+        .grouped(batchSize)
+        .toList
+        .traverse(a => client.assets.updateByExternalId(a.toMap)) *> IO.unit
+    } else {
+      IO.unit
+    }
+
+  def getCdfRootAndTree(sourceRootExternalId: CogniteExternalId): IO[(Option[Asset], List[Asset])] = {
+    val cdfTree = client.assets
+      .filter(AssetsFilter(rootIds = Some(Seq(sourceRootExternalId))))
+      .compile
+      .toList
+
+    for {
+      tree <- cdfTree
+      root = tree.find(_.externalId.contains(sourceRootExternalId.externalId))
+    } yield (root, tree)
+  }
+
+  def upsertRoot(
+      root: AssetsIngestSchema,
+      sourceRoot: Option[Asset]
+  ): IO[Unit] =
+    sourceRoot match {
+      case None =>
+        client.assets.create(
+          Seq(
+            AssetCreate(
```

Comment:
Could we use `AssetsIngestSchema.toAssetCreate` here or is it doing something special?

Reply:
I let `parentExternalId == ""` rather than `null` denote the root Asset, to let the schema have `parentExternalId: String` rather than `parentExternalId: Option[String]`. This handles that by setting `parentExternalId` to `None` in the `AssetCreate`.

Reply:
`AssetsIngestSchema.toAssetCreate(root).copy(parentExternalId = None)`?

Reply:
Yes!